src/librustc_trans/collector.rs

```diff
@@ -433,6 +434,40 @@ fn check_recursion_limit<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
     (instance.def, recursion_depth)
 }
 
+fn check_type_length_limit<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
```

Comment:
Sorry, I wrote my message before seeing yours. Any chance this can be done with recursion, to avoid the memory allocations of the walker? Or is it amortized and otherwise not a problem?
Could we reuse the walker?